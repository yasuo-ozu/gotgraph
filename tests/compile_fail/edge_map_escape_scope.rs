use gotgraph::prelude::*;

fn main() {
    let mut graph: VecGraph<i32, &str> = VecGraph::default();
    
    graph.scope_mut(|mut ctx| {
        let n0 = ctx.add_node(1);
        let n1 = ctx.add_node(2);
        ctx.add_edge("test", n0, n1);
    });
    
    // Try to move edge map outside of scope - should fail
    let escaped_map = graph.scope(|ctx| {
        let edge_map = ctx.init_edge_map(|_, edge| edge.len());
        edge_map // ERROR: edge map cannot escape the scope
    });
    
    // Try to use the escaped map - should fail
    let _ = escaped_map;
}