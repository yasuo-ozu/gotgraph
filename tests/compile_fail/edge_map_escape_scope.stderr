error: lifetime may not live long enough
  --> tests/compile_fail/edge_map_escape_scope.rs:15:9
   |
13 |     let escaped_map = graph.scope(|ctx| {
   |                                    ---- return type of closure `impl Mapping<<gotgraph::graph::Context<'_, &gotgraph::vec_graph::VecGraph<i32, &str>> as gotgraph::graph::Graph>::EdgeIx, usize>` contains a lifetime `'2`
   |                                    |
   |                                    has type `&'1 gotgraph::graph::Context<'_, &gotgraph::vec_graph::VecGraph<i32, &str>>`
14 |         let edge_map = ctx.init_edge_map(|_, edge| edge.len());
15 |         edge_map // ERROR: edge map cannot escape the scope
   |         ^^^^^^^^ returning this value requires that `'1` must outlive `'2`
   |
   = note: requirement occurs because of the type `EdgeTag<'_, EdgeIx>`, which makes the generic argument `'_` invariant
   = note: the struct `EdgeTag<'scope, I>` is invariant over the parameter `'scope`
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance

error: lifetime may not live long enough
  --> tests/compile_fail/edge_map_escape_scope.rs:15:9
   |
13 |     let escaped_map = graph.scope(|ctx| {
   |                                    ---- return type of closure is impl Mapping<<gotgraph::graph::Context<'4, &gotgraph::vec_graph::VecGraph<i32, &str>> as gotgraph::graph::Graph>::EdgeIx, usize>
   |                                    |
   |                                    has type `&gotgraph::graph::Context<'3, &gotgraph::vec_graph::VecGraph<i32, &str>>`
14 |         let edge_map = ctx.init_edge_map(|_, edge| edge.len());
15 |         edge_map // ERROR: edge map cannot escape the scope
   |         ^^^^^^^^ returning this value requires that `'3` must outlive `'4`
